src/theory/uf/theory_uf.h

```diff
@@ -185,31 +188,51 @@ class TheoryUF : public Theory {
   /** called when two equivalence classes are made disequal */
   void eqNotifyDisequal(TNode t1, TNode t2, TNode reason);
 
-private: // for higher-order
-  /** applyExtensionality 
-   * Given disequality deq 
-   * If not already cached, this sends a lemma of the form 
+ private:  // for higher-order
+  /** get extensionality disequality
+   *
+   * Given disequality deq f != g, this returns the disequality:
+   *   (f k) != (g k) for fresh constant(s) k.
+   */
+  Node getExtensionalityDeq(TNode deq);
+
+  /** applyExtensionality
+   *
+   * Given disequality deq f != g, if not already cached, this sends a lemma of
+   * the form:
    *   f = g V (f k) != (g k) for fresh constant k.
-   * on the output channel.
-   * Return value is the number of lemmas sent.
+   * on the output channel. This is a "extensionality lemma".
+   * Return value is the number of lemmas of this form sent on the output
+   * channel.
    */
   unsigned applyExtensionality(TNode deq);
 
-  /** check whether extensionality should be applied for any
-   * pair of terms in the equality engine.
+  /**
+   * Check whether extensionality should be applied for any pair of terms in the
+   * equality engine.
+   *
+   * If we pass a null model m to this function, then we add extensionality
+   * lemmas to the output channel and return the total number of lemmas added.
+   * We only add lemmas for functions whose type is finite.
+   *
+   * If we pass a non-null model m to this function, then we add disequalities
+   * that correspond to the conclusion of extensionality lemmas to the model's
+   * equality engine. We only add disequalities for functions whose type is
+   * infinite. We return 0 if the equality engine of m is consistent after this
+   * call, and 1 otherwise.
```

Comment:
Say why only lemmas for finite or infinite depending on the model being non-null or null.